### ISSUE TITLE
concurrent.futures._base.CancelledError 

### ISSUE THREAD

**Issue by TsioryAntonio** [CJWorkbench/channels_rabbitmq#28]
Hi,

We are using channels with channels-rabbitmq to update our UIs when database entry changes.

In our production environment, we are using gunicorn for django http request and daphne for websocket.

In production mode, when have concurrent.futures._base.CancelledError exception when sending message from a django view, model and signal.

```
only PLAIN login_method is supported, falling back to AMQPLAIN
Unhandled exception from aioamqp
Traceback (most recent call last):
  File "/opt/automate-it/apps/automate-it/ait_factory/vait2/lib/python3.7/site-packages/channels_rabbitmq/connection.py", line 527, in _connect_forever
    await self._connect_and_run()
  File "/opt/automate-it/apps/automate-it/ait_factory/vait2/lib/python3.7/site-packages/channels_rabbitmq/connection.py", line 592, in _connect_and_run
    await self._channel.close_event.wait()  # case 1, 2, 3
  File "/usr/local/lib/python3.7/asyncio/locks.py", line 293, in wait
    await fut
concurrent.futures._base.CancelledError
error on dispatch
Traceback (most recent call last):
  File "/opt/automate-it/apps/automate-it/ait_factory/vait2/lib/python3.7/site-packages/aioamqp/protocol.py", line 316, in run
    await self.dispatch_frame()
  File "/opt/automate-it/apps/automate-it/ait_factory/vait2/lib/python3.7/site-packages/aioamqp/protocol.py", line 264, in dispatch_frame
    frame_channel, frame = await self.get_frame()
  File "/opt/automate-it/apps/automate-it/ait_factory/vait2/lib/python3.7/site-packages/aioamqp/protocol.py", line 251, in get_frame
    return await amqp_frame.read(self._stream_reader)
  File "/opt/automate-it/apps/automate-it/ait_factory/vait2/lib/python3.7/site-packages/aioamqp/frame.py", line 78, in read
    data = await reader.readexactly(7)
  File "/usr/local/lib/python3.7/asyncio/streams.py", line 679, in readexactly
    await self._wait_for_data('readexactly')
  File "/usr/local/lib/python3.7/asyncio/streams.py", line 473, in _wait_for_data
    await self._waiter
concurrent.futures._base.CancelledError
```

Can you help us to deal with this issue? 

Here is our environment:

```
asgiref (3.2.10)
channels (2.4.0)
channels-rabbitmq (1.2.1)
Django (2.2.16)  
gunicorn (20.0.4)
```

Here is our consumer: 

```
class SequenceConsumer(WebsocketConsumer):
    GROUP_NAME = "AIT_Sequence"

    def connect(self):
        user = self.scope["user"]

        if user and user.id:
            async_to_sync(self.channel_layer.group_add)(
                self.GROUP_NAME,
                self.channel_name
            )
            self.accept()

    def disconnect(self, close_code):
        async_to_sync(self.channel_layer.group_discard)(
            self.GROUP_NAME,
            self.channel_name
        )

    def message_send(self, event):
        self.send(text_data=event["text"])
```
        
In our views, our models and our signals we use:
```
SequenceNotification.send_notification({"instance": sequence, "message": message, 'bypass_perm': True})

class SequenceNotification(object):
    required_perms = ["view_ait_sequencetemplate"]
    model_type = 'AIT_Sequence'
    channel_layer = get_channel_layer()


    @classmethod
    def send_notification(cls, notification):
        async_to_sync(cls.channel_layer.group_send)(
            cls.model_type, {
                "type": "message.send",
                "text": json.dumps(notification['message'])
            }
        )
```

**Comment by adamhooper** [comment]
Does the problem go away if you make it an `AsyncWebsocketConsumer`?

My hunch is that `async_to_sync()` isn't finding an existing event loop -- so you're getting a new event loop with each call. That forces a brand-new RabbitMQ connect+disconnect each call. When the event loop shuts down, `CancelledError` is raised on the connection's background task. But even if it _weren't_ raised, you'd have a bigger problem: the disconnect forces `group_discard()`.

**Comment by TsioryAntonio** [comment]
Thanks for quick reply.

The problem doesn't go away when using AsyncWebsocketConsumer. 

From where does the event loop come from? Does it exist in a django view or model context (which is run by gunicorn as WSGI server)?

**Comment by adamhooper** [comment]
What command are you using to start your web server?

**Comment by TsioryAntonio** [comment]
This one for gunicorn service:
`gunicorn ait_factory.wsgi:application --env DJANGO_SETTINGS_MODULE=ait_factory.settings.factory --bind=0.0.0.0:8000`

And this one for daphne:
`daphne -b 0.0.0.0 -p 8001 ait_factory.asgi:application`

**Comment by adamhooper** [comment]
A-ha. This is exactly the sort of mistake that `async_to_sync()` encourages. And _waaaaay_ too many docs let this happen. And `channels_redis` supports the antipattern! `channels_redis` will connect+disconnect just to send the one message, but it will give your expected behavior because it doesn't _listen_ for messages to the server. (`channels_redis` _polls_ for messages with a request+response pattern. `channels_rabbitmq` _listens_ for messages in the background -- and this background task is being cancelled incorrectly.)

In a nutshell: `async_to_sync()` should only apply when you call it _from within sync_to_async()_.

Daphne meets that condition: it wraps every Django view call in `sync_to_async()`, so calling `async_to_sync()` from within Daphne-served Django views is fine.

Gunicorn doesn't meet this condition: it doesn't run within asyncio, and so each call to `async_to_sync()` spawns a new event loop. This isn't what you want. gunicorn + Channels (or at least, `channels_rabbitmq`) isn't suported.

Here are your paths forward:

1. Stop using gunicorn: use Daphne or uvicorn instead.
2. Write an async function that connects to RabbitMQ, sends the message, and then disconnects; and call _that_ method within `async_to_sync()` in your gunicorn views. (You don't need `channels_rabbitmq` for this: you can use any async RabbitMQ client.)
3. Write a sync function that connects to RabbitMQ, sends the message, then disconnects. Call this from your gunicorn views. (You can use any sync RabbitMQ client.)
4. Write a (sync) background thread that connects to RabbitMQ, listens for messages on a queue, and sends each message to RabbitMQ. (This wouldn't be `channels_rabbitmq` either.)

It's _depressing_. And I can't help but rant. If you've already committed to `channels_redis` and you're fed up with its leaks and lock-ups, then yes, use `channels_rabbitmq` ... but only within WebsocketConsumers. (Use other libraries to fill the remaining gaps.)

If you're adding websockets to an existing sync app that works the way you want, stop. Two years into it, you'll have learned how awful Python asyncio is (see this error message for one example); plus you'll encounter many bad design decisions in Channels. You'll become _shocked_ that the Channels website _only_ recommends `channels_redis`, which locks up by design and drops connections by design. It's a _lot_ of effort keeping up with the mess. And this mess is unique to the Python community. You might be better off writing your Websockets server in NodeJS.

**Comment by adamhooper** [comment]
Closed in v2.0: now, we test `async_to_sync()`.

All the provisos in https://github.com/CJWorkbench/channels_rabbitmq/issues/28#issuecomment-734334065 still apply. You _shouldn't_ use `async_to_sync()` outside of Django-Channels `SyncConsumer` or Django 3.1 views with async middleware.

In particular, don't use `async_to_sync()` in Celery task workers or in HTTP handlers that _aren't_ served from an ASGI application. (Daphne = good; uvicorn = good; gunicorn = bad.)

`channels_rabbitmq` shouldn't crash in such a case. If it does, that's a bug in `channels_rabbitmq`. But misusing `async_to_sync()` in this manner near-guarantees you'll face serious, hard-to-fix problems down the road.

**Comment by adamhooper** [comment]
Updated in v3.0:

Now, `async_to_sync()` outside of `sync_to_async()` will raise a RuntimeError. It's simpler this way.

(If you'd like the previous hack restored, please file a new issue report. The description should begin with a paraphrased version of the following: _I understand why calling async_to_sync() outside of sync_to_async() forces a new RabbitMQ connection per call; I know this isn't documented anywhere in the Django Channels documentation; I know that there is no documented way to shut down a channel layer with every call; and *still* allowing this pattern is appropriate to my needs. Here's why: ...._)

On the bright side, I'm pleased to announce [Carehare](https://github.com/CJWorkbench/carehare), our very own RabbitMQ client. It's designed to make debugging easy. Here's a snippet you can use in `async_to_sync()`:

```
import asyncio
from typing import Any, Dict

import carehare
import msgpack
from django.conf import settings

def group_send_sync(group: str, message: Dict[str, Any]) -> None:
    """Synchronously connect to RabbitMQ, send a message to a group, and disconnect.

    Raise OSError, carehare.ConnectionClosedByServer or carehare.ChannelClosedByServer
    on error.
    """
    async def group_send():
        async with carehare.connect(
            url=settings.CHANNEL_LAYERS["default"]["CONFIG"]["host"]
        ) as connection:
            await connection.publish(
                msgpack.packb({"__asgi_group__": group, **message}),
                exchange_name="groups",
                routing_key=group,
            )
    asyncio.run(group_send())
```
